clients/spark/core/src/main/scala/io/treeverse/clients/GarbageCollector.scala

```diff
@@ -0,0 +1,244 @@
+          getSeconds(a.message.lastModified)
+        )
+      )
+      .toSet
+  }
+  def getAddressesTuples(
+      rangeIDs: Set[String],
+      conf: APIConfigurations
+  ): Set[(String, String, Boolean, Long)] = {
+    if (rangeIDs.isEmpty) Set()
+    else
+      getEntryTuples(rangeIDs.head, conf).union(getAddressesTuples(rangeIDs.tail, conf))
+  }
+
+  def subtractAddressesInRange(
+      addresses: Set[(String, String, Boolean, Long)],
+      rangeID: String,
+      conf: APIConfigurations,
+      repo: String
+  ): Set[(String, String, Boolean, Long)] = {
+    val activeAddresses = getRangeAddresses(rangeID, conf, repo)
+    addresses.filterNot(x => activeAddresses.contains(x._1))
+  }
+
+  /** @param leftRangeIDs
+   *  @param rightRangeIDs
+   *  @param conf
+   *  @return tuples of type (key, address, isRelative, lastModified) for every address existing in leftRanges and not in rightRanges
+   */
+  def leftAntiJoinAddresses(
+      leftRangeIDs: Set[String],
+      rightRangeIDs: Set[String],
+      conf: APIConfigurations,
+      repo: String
+  ): Set[(String, String, Boolean, Long)] = {
+    @tailrec
+    def subtractAddressesInRanges(
+        addresses: Set[(String, String, Boolean, Long)],
+        rangeIDs: Set[String]
+    ): Set[(String, String, Boolean, Long)] = {
+      if (rangeIDs.isEmpty) addresses
+      else
+        subtractAddressesInRanges(
+          subtractAddressesInRange(addresses, rangeIDs.head, conf, repo),
+          rangeIDs.tail
+        )
+    }
+    subtractAddressesInRanges(getAddressesTuples(leftRangeIDs, conf), rightRangeIDs)
+  }
+
+  /** receives a dataframe containing active and expired ranges and returns addresses contained only in expired ranges
+   *  @param ranges dataframe of type   rangeID:String | expired: Boolean
+   *  @return dataframe of type  key:String | address:String | relative:Boolean | last_modified:Long
+   */
+  def getAddressesDFFromRanges(
+      ranges: Dataset[Row],
+      conf: APIConfigurations,
+      repo: String
+  ): Dataset[Row] = {
+    val left_anti_join_addresses = udf((x: Seq[String], y: Seq[String]) => {
+      leftAntiJoinAddresses(x.toSet, y.toSet, conf, repo).toSeq
+    })
+    val expiredRangesDF = ranges.where("expired")
+    val activeRangesDF = ranges.where("!expired")
+
+    // ranges existing in expired and not in active
+    val uniqueExpiredRangesDF = expiredRangesDF.join(
+      activeRangesDF,
+      expiredRangesDF("range_id") === activeRangesDF("range_id"),
+      "leftanti"
+    )
+
+    // for every expired range - get all intersecting active ranges
+    //  expired_range | active_ranges  | min-max (of expired_range)
+    val joinActiveByRange = uniqueExpiredRangesDF
+      .as("u")
+      .join(
+        activeRangesDF.as("a"),
+        (col("a.max_key") <= col("u.max_key")) && (col("a.max_key") >= col("u.min_key")) ||
+          ((col("a.min_key") <= col("u.max_key")) && (col("a.min_Key") >= col("u.min_key"))),
```

Comment:
I find it easier to read "between" checks that look more like the mathematical formulation "a <= b <= c", and also many fewer brackets, e.g.
```suggestion
        col("u.min_key") <= col("a.max_key") && col("a.max_key") <= col("u.max_key") ||
        col("u.min_key") <= col("a.min_key") && col("a.min_key") <= col("u.max_key")),
```

If you do need those extra brackets for clarity, I would place them one level up, i.e. replacing "A && B || C && D" above with "(A && B) || (C && D)" (the original code is "(A) && (B) || (C) && (D)", which I feel only adds noise).